src/components/viewmodels/rooms/UserIdentityWarningViewModel.tsx

```diff
@@ -0,0 +1,164 @@
+/*
+Copyright 2025 New Vector Ltd.
+
+SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
+Please see LICENSE files in the repository root for full details.
+*/
+
+import { useCallback, useEffect, useMemo, useState } from "react";
+import { EventType, MatrixEvent, Room, RoomStateEvent, RoomMember } from "matrix-js-sdk/src/matrix";
+import { CryptoEvent, CryptoApi } from "matrix-js-sdk/src/crypto-api";
+import { throttle } from "lodash";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import { useMatrixClientContext } from "../../../contexts/MatrixClientContext.tsx";
+import { useTypedEventEmitter } from "../../../hooks/useEventEmitter.ts";
+import { ButtonEvent } from "../../views/elements/AccessibleButton.tsx";
+
+export type ViolationType = "PinViolation" | "VerificationViolation";
+
+export type ViolationPrompt = {
+    member: RoomMember;
+    type: ViolationType;
+};
+
+export interface UserIdentityWarningState {
+    currentPrompt?: ViolationPrompt;
+    onButtonClick: (ev: ButtonEvent) => void;
+}
+
+async function mapToViolations(cryptoApi: CryptoApi, members: RoomMember[]): Promise<ViolationPrompt[]> {
```

Comment:
This could also do with some doc as it's not super obvious what it does.

Reply:
Done https://github.com/element-hq/element-web/pull/29067/commits/6b76fbc39273ebef8366a2eb3492e6d87d8f9c67